server/resp/src/main/java/org/infinispan/server/resp/commands/generic/SCAN.java

```diff
@@ -0,0 +1,140 @@
+               } else {
+                  try {
+                     count = Integer.parseInt(new String(arguments.get(i), StandardCharsets.US_ASCII));
+                  } catch (NumberFormatException e) {
+                     Consumers.VALUE_NOT_INTEGER_BICONSUMER.accept(null, handler.allocatorToUse());
+                     return handler.myStage();
+                  }
+               }
+            } else if (Arrays.compare(TYPE, arg) == 0) {
+               if (++i >= argc) {
+                  Consumers.SYNTAX_ERROR_BICONSUMER.accept(null, handler.allocatorToUse());
+                  return handler.myStage();
+               } else {
+                  try {
+                     type = RespTypes.valueOf(new String(arguments.get(i), StandardCharsets.US_ASCII));
+                  } catch (IllegalArgumentException e) {
+                     type = RespTypes.unknown;
+                  }
+               }
+            }
+         }
+      }
+
+      DefaultIterationManager iterationManager = handler.respServer().getIterationManager();
+      String cursor = new String(arguments.get(0), StandardCharsets.US_ASCII);
+      if ("0".equals(cursor)) {
+         // New iterator
+         if (count < 0) {
+            count = DEFAULT_COUNT;
+         }
+         IterationState iterationState = iterationManager.start(handler.cache(), null, filterConverterFactory, filterConverterParams, MediaType.APPLICATION_OCTET_STREAM, count, false, DeliveryGuarantee.EXACTLY_ONCE);
+         iterationState.getReaper().registerChannel(ctx.channel());
+         cursor = iterationState.getId();
+      }
+      IterableIterationResult result = iterationManager.next(cursor, count);
+      if (result.getStatusCode() == IterableIterationResult.Status.InvalidIteration) {
+         // Let's just return 0
+         ByteBufferUtils.stringToByteBuf("*2\r\n$1\r\n0\r\n*0\r\n", handler.allocatorToUse());
+      } else {
+         StringBuilder response = new StringBuilder();
+         response.append("*2\r\n");
+         int size = result.getEntries().size();
+         if (size < count) {
+            // We've reached the end of iteration, return a 0 cursor
+            response.append("$1\r\n0\r\n");
+            iterationManager.close(cursor);
+         } else {
+            response.append('$');
+            response.append(cursor.length());
+            response.append(CRLF);
+            response.append(cursor);
+            response.append(CRLF);
+         }
+         response.append('*');
+         response.append(size);
+         response.append(CRLF);
+         for(CacheEntry entry : result.getEntries()) {
+            byte[] key = (byte[]) entry.getKey();
+            response.append('$');
+            response.append(key.length);
+            response.append(CRLF);
+            response.append(new String(key));
+            response.append(CRLF);
+         }
+         ByteBufferUtils.stringToByteBuf(response.toString(), handler.allocatorToUse());
```

Comment:
We can avoid this String allocation with something like:

```suggestion
         List<byte[]> keys = new ArrayList<>(size);
         for(CacheEntry entry : result.getEntries()) {
            keys.add((byte[]) entry.getKey());
         }

         ByteBufferUtils.stringToByteBuf(response, handler.allocatorToUse());
         ByteBufferUtils.bytesToResult(keys, handler.allocatorToUse());
```